allennlp/training/optimizers.py

```diff
@@ -26,13 +27,50 @@ class Optimizer(Registrable):
     default_implementation = "adam"
 
     @classmethod
-    def from_params(cls, model_parameters: List[torch.nn.Parameter], params: Params):
+    def from_params(cls, model_parameters: List, params: Params):
         if isinstance(params, str):
             optimizer = params
             params = Params({})
         else:
             optimizer = params.pop_choice("type", Optimizer.list_available())
-        return Optimizer.by_name(optimizer)(model_parameters, **params.as_dict()) # type: ignore
+
+        # make the parameter groups if need
+        groups = params.pop("parameter_groups", None)
+        if groups:
+            # input to optimizer is list of dict
+            # each dict contains {'params': [list of parameters], 'lr': 1e-3, ...}
+            # see: http://pytorch.org/docs/0.3.0/optim.html?#per-parameter-options
+            #
+            # groups contains something like:
+            #"parameter_groups": [
+            #       [['regex1', 'regex2'], {'lr': 1e-3},
+            #        ['regex3'], {'lr': 1e-4}]
+            #]
+            # last entry is for the parameters not in any regex
+            parameter_groups: Any = [{'params': []} for _ in range(len(groups) + 1)]
```

Comment:
I think this type should be `List[Dict[str, List[torch.nn.Parameter]]]`. If there was a reason it had to be `Any`, can you say why (using `Any` as a type removes all of the benefit we get from using MyPy)

Reply:
All of the key/value pairs except `params` are used to set the kwargs for the parameter group (`lr`, etc), so that the type of the `Dict` values can be anything accepted by the optimizer constructor.  This is done in the statement `parameter_groups[k].update(groups[k][1].as_dict())` a few lines down.  So I used the `Any` type to avoid enumerating all possible types.

Reply:
Ah gotcha.